### DISCUSSION TITLE
Concatenade StringRef and StringLiteral

### DISCUSSION THREAD

**Discussion by gardberg** [modular/modular#1059] [Q&A]
Hi, I want to concatenate two string variables into a single one like this:

```
let API_KEY = getenv("API_KEY") # StringRef
let prefix = "Bearer " # StringLiteral

var content = Python.dict()
content["Content-Type"] = "application/json"
content["Authorization"] = prefix + API_KEY # <-- Not allowed

# Use content in a post request
...
```

The compiler says:

invalid call to '__add__': right side cannot be converted from 'StringRef' to 'StringLiteral'

What is the proper way to concatenate these strings? What are the differences between StringRef and StringLiteral types?

**Comment by rd4com** [comment]
Hello, i don't know if it is the proper way:
```python
fn main():
    var original:String = "afterprefix"
    var prefix: StringLiteral = "prefix_"
    var original_ref: StringRef = StringRef(original._as_ptr(),len(original))
     
    # note that the string literal is explicitly converted to a string   
    print(String(prefix)+original_ref)
    
    #extend the lifetime of original. 
    #if not, stringref will point to freed memory
    #see "asap" policy (as soon as possible) in the docs
    _=original
```
explicit conversion: ``String(prefix)+String(original_ref)``

Nice to see you are working on cool stuff!

**Comment by Moosems** [comment]
Yes, I'd assume that converting StringLiterals to Strings is the way to go here (and honestly in most cases).
